docs/sources/lids/_DRAFT-Index-Gateway-Sharding.md

```diff
@@ -0,0 +1,97 @@
+
+## Problem Statement
+
+The current strategy of sharding by tenant ID and having a replication factor fails in the long run, because even when running lots of index gateways, only a maximum of `n` instances could be utilized by a single tenant, where `n` is the value of the configured RF.
+
+Another problem is that the RF is fixed and the same for all tenants, independent of their actual size in terms of log volume or query rate.
+
+## Goals
+
+The goal of this document is to find a better sharding mechanism for the index gateway, so that there are no boundaries for scaling the service horizontally.
+
+* The sharding needs to account for the "size" of a tenant.
+* A single tenant needs to be able to utilize more than three index gateways.
+
+## Proposals
+
+### Proposal 0: Do nothing
+
+If we do not improve the sharding mechanism for the index gateways and leave it as it is, it will become more and more difficult to serve metadata queries for large tenant in a reasonable amount of time, proportionally to the demand for these queries.
+
+### Proposal 1: Dynamic replication factor
+
+Instead of using a fixed replication factor of 3, the RF can be derived from the amount of active members in the index gateway ring. That means that the RF would be a percentage of the available gateway instances. For example, a ring with 12 instances and 30% replication utilization would result in a RF of 3 (`floor(12*0.3)`). Scaling up to 18 instances would result in a RF of 5.
+
+This approach would solve the problem of horizontal scaling. However, it does not solve the problem of different tenant sizes. It also fails to ensure replication for availability when running a small number of instances, unless there is a fixed lower value for the RF. It also tends to over-shard data in large index gateway deployments.
+
+### Proposal 2: Fixed per-tenant replication factor
+
+Adding a random shard ID (e.g. `shard-0`, `shard-1`, ... `shard-n`) to the tenant ID allows to utilize a certain amount of `n` instances. The amount of shards can be implemented as a per-tenant override setting. This would allow to use different amount of instances for each tenant. However, this approach results in non-deterministic hash keys.
+
+### Proposal 3: Shard by index files
+
+In order to answer requests, the index gateway needs to download index files from the object storage, and since Loki builds a daily index file per tenant, these index files can be sharded evenly across all available index gateway instances. Each instance is then assigned a unique set of index files which it can answer metadata queries for.
+
+This means that the sharding key is the name of the file in object storage. While this name encodes both the tenant and the date, this is not strictly necessary. Such a sharding mechanism could shard any files from object storage across a set of instances of a ring.
```

Comment:
Sharding by file name solves the problem of distributing a single tenant's files across multiple instances. However, it does not solve the problem of sharding the workload of a single tenant across, because metadata queries for a single day are assigned to a single instance regardless of how frequent the data is queried and how big the index is.

Reply:
This could be solved by tracking the access rate to a single resource (index file) and based on that add a replica suffix to the filename for the hash function input.

Dummy example:

```
   RPS -> MAX_REPLICAS
  1rps -> 1
 10rps -> 2
100rps -> 4

key = HASH_FUNC(FILENAME + ";" + RAND_INT(0, MAX_REPLICAS))
```

This would not shard a single resource that is requested more often to multiple instances, essentially spreading the load of the requests, when every request picks a random replica.

Note, that this replica is independent of the ring's replication factor!